root/static/scripts/common/components/EntityLink.js

```diff
@@ -32,96 +36,109 @@ const DeletedLink = ({name, allowNew}) => {
+  const dates = formatDatePeriod(event);
+  if (!dates && !event.cancelled) {
+    return null;
   }
-}
-
-class AreaDisambiguation extends React.Component {
-  render() {
-    let area = this.props.area;
-
-    if (!area.ended) {
-      return null;
-    }
+  return (
+    <Frag>
+      {dates ? bracketed(dates) : null}
+      {event.cancelled
+        ? <Comment className="cancelled" comment={l('cancelled')} />
+        : null}
+    </Frag>
+  );
+};
 
-    let comment;
-    let beginYear = area.begin_date ? area.begin_date.year : null;
-    let endYear = area.end_date ? area.end_date.year : null;
+const AreaDisambiguation = ({area}: {|+area: AreaT|}) => {
+  if (!area.ended) {
+    return null;
+  }
 
-    if (beginYear && endYear) {
-      comment = l('historical, {begin}-{end}', {begin: beginYear, end: endYear});
-    } else if (endYear) {
-      comment = l('historical, until {end}', {end: endYear});
-    } else {
-      comment = l('historical');
-    }
+  let comment;
+  const beginYear = area.begin_date ? area.begin_date.year : null;
+  const endYear = area.end_date ? area.end_date.year : null;
 
-    return <Comment className="historical" comment={comment} />;
+  if (beginYear && endYear) {
+    comment = l('historical, {begin}-{end}', {begin: beginYear, end: endYear});
+  } else if (endYear) {
+    comment = l('historical, until {end}', {end: endYear});
+  } else {
+    comment = l('historical');
   }
-}
+
+  return <Comment className="historical" comment={comment} />;
+};
 
 const NoInfoURL = ({url, allowNew}) => (
   <Frag>
     <a href={url}>{url}</a>
     {' '}
-    <DeletedLink name={'[' + l('info') + ']'} allowNew={allowNew} />
+    <DeletedLink allowNew={allowNew} name={'[' + l('info') + ']'} />
   </Frag>
 );
 
-const EntityLink = (props = {}) => {
-  let {
-    allowNew,
-    content,
-    entity,
-    hover,
-    showDeleted = true,
-    showDisambiguation,
-    subPath,
-    ...anchorProps
-  } = props;
-
+/* eslint-disable sort-keys, flowtype/sort-keys */
+type EntityLinkProps = {
+  +allowNew?: boolean,
+  +content?: React.Node,
+  +entity: CoreEntityT,
+  +hover?: string,
+  +showDeleted?: boolean,
+  +showDisambiguation?: boolean,
+  +subPath?: string,
+
+  // ...anchorProps
+  href?: string,
+  title?: string,
+  +target?: '_blank',
+};
+/* eslint-enable sort-keys, flowtype/sort-keys */
+
+const EntityLink = ({
+  allowNew = false,
+  content,
+  entity,
+  hover,
+  showDeleted = true,
+  showDisambiguation,
+  subPath,
+  ...anchorProps
+}: EntityLinkProps) => {
   const hasCustomContent = nonEmpty(content);
-  const entityType = entity.entityType;
-  const comment = ko.unwrap(entity.comment);
+  const comment = entity.comment ? ko.unwrap(entity.comment) : '';
 
   if (showDisambiguation === undefined) {
     showDisambiguation = !hasCustomContent;
   }
 
-  if (entityType === 'artist' && !nonEmpty(hover)) {
+  if (entity.entityType === 'artist' && !nonEmpty(hover)) {
```

Comment:
Since `CoreEntityT` is now a [disjoint union](https://flow.org/en/docs/types/unions/#toc-disjoint-unions), the `entityType` check allows Flow to determine that the entity must be an `ArtistT` inside this block, allowing the `entity.sort_name` access.

But for some reason this feature only works if you access `entity.entityType` directly in the conditional, rather than caching it in a variable as we were doing before.

Reply:
Is that limitation documented or reported?

Reply:
I found https://github.com/facebook/flow/issues/3932 and https://github.com/facebook/flow/issues/4828 (though they're related to destructuring the values more specifically, I guess the same logic would apply).

It might also be related to [refinement invalidation](https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations) in that Flow can't know if `entity.entityType` has changed value since it was stored in the constant...so it may be a way to enforce soundness.